### PR TITLE
Add more detail to README's authentication section

### LINKED ISSUE #331: We should mention GOOGLE_APPLICATION_CREDENTIALS in the Authentication section.
The main gcloud-java [README.md](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/README.md) contains an authentication section however this section does not
mention the option of using the `GOOGLE_APPLICATION_CREDENTIALS` environment variable to point to a key file.

@mziccard I now wonder if we need the `GCLOUD_TESTS_KEY` and `GCLOUD_TESTS_PROJECT_ID` in the `RemoteGcsHelper` as `GOOGLE_APPLICATION_CREDENTIALS` could be used instead of the former and `GCLOUD_PROJECT` could be used instead of the latter.

### PR DESCRIPTION
**Pull request by ajkannan**
Fixes #331

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,9 +44,19 @@ There are multiple ways to authenticate to use Google Cloud services.
 
 1. When using `gcloud-java` libraries from within Compute/App Engine, no additional authentication steps are necessary.
 2. When using `gcloud-java` libraries elsewhere, there are two options:
-  * [Generate a JSON service account key](https://cloud.google.com/storage/docs/authentication?hl=en#service_accounts).  Supply a path to the downloaded JSON credentials file when building the options supplied to datastore/storage constructor.
+  * [Generate a JSON service account key](https://cloud.google.com/storage/docs/authentication?hl=en#service_accounts).  After downloading that key, you must do one of the following:
+    * Define the environment variable GOOGLE_APPLICATION_CREDENTIALS to be the location of the key.  For example, `export GOOGLE_APPLICATION_CREDENTIALS=/path/to/my/key.json`
+    * Supply the downloaded JSON credentials file when building the options supplied to datastore/storage constructor.  For example, `StorageOptions.builder().authCredentials(AuthCredentials.createForJson(new FileInputStream("/path/to/my/key.json")).build().service()` returns a `Storage` object that has the necessary permissions.
```
Comment (aozarov) [inline README.md]: when building the service options.

Comment (aozarov) [inline README.md]: Example is too long. Maybe should break the lines like this:

* Define the environment variable GOOGLE_APPLICATION_CREDENTIALS to be the location of the key.  For example:
    ```bash
    export GOOGLE_APPLICATION_CREDENTIALS=/path/to/my/key.json
    ```

* Supply the downloaded JSON credentials file when building the options supplied to datastore/storage constructor.  For example:
    ```java
    StorageOptions.builder().authCredentials(
        AuthCredentials.createForJson(new FileInputStream("/path/to/my/key.json")).build().service()
    ``` 
returns a `Storage` object that has the necessary permissions.

File: README.md [hunk 2]
```diff
@@ -44,9 +44,19 @@ There are multiple ways to authenticate to use Google Cloud services.
...
   * If running locally for development/testing, you can use use [Google Cloud SDK](https://cloud.google.com/sdk/?hl=en).  To use the SDK authentication, [download the SDK](https://cloud.google.com/sdk/?hl=en) if you haven't already.  Then login using the SDK (`gcloud auth login` in command line), and set your current project using `gcloud config set project PROJECT_ID`.
```
Comment (aozarov) [inline README.md]: As discussed. Maybe we should have a separate section for project id settings, defaults (and order).

Comment (ajkannan) [inline README.md]: Ah, just realized you were only part way through the review.  I pushed some changes, but I'll wait and fix the PR to incorporate these suggestions once you're done.

File: README.md [hunk 3]
```diff
@@ -44,9 +44,19 @@ There are multiple ways to authenticate to use Google Cloud services.
...
 
+`gcloud-java` looks for credentials in the following order, stopping once it finds credentials:
+
+1. Credentials supplied to the `DatastoreOptions`/`ServiceOptions` builder
+2. App Engine credentials
+3. Key file pointed to by the GOOGLE_APPLICATION_CREDENTIALS environment variable
+4. Google SDK credentials
+5. Compute Engine credentials
```
Comment (aozarov) [inline README.md]: I am actually not sure about the order of [4] and [5], though in practice that is not going to be a problem.

`AuthCredentials.createApplicationDefaults()` which was added when #21 was fixed
is going to use CE credentials and also consider SDK credentials (it still has a [bug](https://github.com/google/google-auth-library-java/issues/2) for not consider AE crendetials) but I am not sure in which order. Can you verify?

File: README.md [hunk 4]
```diff
@@ -44,9 +44,19 @@ There are multiple ways to authenticate to use Google Cloud services.
...
+  * [Generate a JSON service account key](https://cloud.google.com/storage/docs/authentication?hl=en#service_accounts).  After downloading that key, you must do one of the following:
+    * Define the environment variable GOOGLE_APPLICATION_CREDENTIALS to be the location of the key.  For example, `export GOOGLE_APPLICATION_CREDENTIALS=/path/to/my/key.json`
+    * Supply the downloaded JSON credentials file when building the options supplied to datastore/storage constructor.  For example, `StorageOptions.builder().authCredentials(AuthCredentials.createForJson(new FileInputStream("/path/to/my/key.json")).build().service()` returns a `Storage` object that has the necessary permissions.
   * If running locally for development/testing, you can use use [Google Cloud SDK](https://cloud.google.com/sdk/?hl=en).  To use the SDK authentication, [download the SDK](https://cloud.google.com/sdk/?hl=en) if you haven't already.  Then login using the SDK (`gcloud auth login` in command line), and set your current project using `gcloud config set project PROJECT_ID`.
```
Comment (aozarov) [inline README.md]: I was actualy thinking of having a different section (in the level of authentication) that describe how to set a project id (and have the "set your current project using `gcloud config set project PROJECT_ID`." suggestion be part of it).

Another option is to have it as a subsection of Authentication but still with a head and a parallel suggestion (prod, dev and order).

File: README.md (hunk 3 above)
Comment (ajkannan) [inline README.md]: Yup, I verified the order of 4 & 5 [here](https://developers.google.com/identity/protocols/application-default-credentials)

**Comment by ajkannan** [comment]
@aozarov Fixed the docs, ready for review.

File: README.md [hunk 5]
```diff
@@ -37,6 +37,36 @@ Example Applications
...
+2. When using `gcloud-java` elsewhere, you can do one of the following:
+  * Supply the project ID when building the service options.  For example, to use Datastore from a project with ID "PROJECT_ID", you can write:
+
+  ```java
+  Datastore datastore = DatastoreOptions.builder().projectId("PROJECT_ID").build().service(); 
+  ```
+  * Specify the environment variable `GCLOUD_PROJECT`.  For example, type the following into command line:
+
+  ```bash
+  export GCLOUD_PROJECT=PROJECT_ID
```
Comment (aozarov) [inline README.md]: I don't think this "example" is necessary.

File: README.md [hunk 6]
```diff
@@ -37,6 +37,36 @@ Example Applications
...
+  ```
+  gcloud config set project PROJECT_ID
+  ```
+
+`gcloud-java` determines the project ID from the following sources in the listed order, stopping once it finds a value:
+
+1. Project ID supplied when building the service options.
+2. Project ID specified by the environment variable `GCLOUD_PROJECT`.
+3. Project ID used by App Engine.
+4. Project ID specified in the Google Cloud SDK.
```
Comment (aozarov) [inline README.md]: see ServiceOptions#googleCloudProjectId it looks like we first try to get it from CE metadata.
Maybe, we should change the order in this function to be consistent with the auth lookup order...

File: README.md [hunk 7]
```diff
@@ -44,8 +74,27 @@ There are multiple ways to authenticate to use Google Cloud services.
...
+    export GOOGLE_APPLICATION_CREDENTIALS=/path/to/my/key.json
+    ```
+    * Supply the JSON credentials file when building the service options.  For example, this Storage object has the necessary permissions to interact with your Google Cloud Storage data:
+    ```java
+    Storage storage = StorageOptions.builder()
+      .authCredentials(AuthCredentials.createForJson(new FileInputStream("/path/to/my/key.json"))
+      .build()
+      .service();
+    ```
+  * If running locally for development/testing, you can use use Google Cloud SDK.  Download the SDK if you haven't already, then login using the SDK (`gcloud auth login` in command line).  Be sure your current project is set correctly by running `gcloud config set project PROJECT_ID`.
```
Comment (aozarov) [inline README.md]: Be sure to set your project ID as described above (or something like that).

File: README.md (hunk 6 above)
Comment (ajkannan) [inline README.md]: Good catch, I'll add CE to the list.

I agree that we should try to be consistent with the auth lookup order.  I vote we keep the order as is in this PR and update the project ID code & README once the auth updates are cleaned up and tested in #334.  I can add this todo item to issue #334.

**Pull request merged**